lwt/util.ml

```diff
@@ -0,0 +1,34 @@
+open Xen_api
+open Lwt
+
+let sockaddr_of_uri uri =
+  let use_ssl =
+    match Uri.scheme uri with
+    | Some "https" ->
+      true
+    | Some "http" ->
+      false
+    | x ->
+      failwith (Printf.sprintf "Unsupported URI scheme: %s" (match x with None -> "None" | Some x -> x))
+  in
+  let port =
+    match Uri.port uri with
+    | None ->
+      (if use_ssl then 443 else 80)
+    | Some port ->
+      port
+  in
+  let host =
+    match Uri.host uri with
+    | None ->
+      failwith "Please supply a host in the URI"
+    | Some host ->
+      Scanf.ksscanf host (fun _ _ -> host) "[%s@]" Fun.id
+  in
+  Lwt.catch (fun () ->
+    Lwt_unix.getaddrinfo host (string_of_int port) [] >|= function
+      | [] ->
+        None
+      | addrinfo :: _ ->
+        Some addrinfo.Unix.ai_addr)
+    (fun _ ->  fail (Failed_to_resolve_hostname host))
```

Comment:
this function returns an option, yet you are throwing exceptions when the `uri` is invalid. It'd be better to pick just one method of error handling, rather than using both

Reply:
So I should return None as well?

Reply:
personally I would throw exceptions, as this seems to be the style of the existing code